### ISSUE TITLE
disallowed_methods falsely claiming "does not refer to a reachable function" (workspaces?)

### ISSUE THREAD

**Issue by ijackson** [rust-lang/rust-clippy#17006]
### Summary

In our codebase, we see the warning "does not refer to a reachable function" for one of the methods we have disabled in our `clippy.toml` using `disallowed_methods`.

I suspect that this is because the method is in a crate that is not a dependency of every crate in our worksapce?  Or something.

### Lint Name

disallowed_methods

### Reproducer

# Steps

```
git clone https://gitlab.torproject.org/tpo/core/arti
cd arti
git checkout f7b6a9d27408d07fafce9c90530177abe7b751ab
cargo clippy --locked --all-features --workspace --all-targets
```

# Expected results

No warnings.

# Actual results

```
warning: `rand::RngExt::random_range` does not refer to a reachable function
 --> /volatile/rustcargo/d/arti/clippy.toml:7:5
  |
7 |     { path = "rand::RngExt::random_range", reason = "Prefer tor_basic_utils::RngExt::gen_range_checked to avoid uncontrolled panics, or gen_range_infallible if applicable" },
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: add `allow-invalid = true` to the entry to suppress this warning
 
warning: `tor-geoip` (lib test) generated 1 warning
```

### Version

```text
rustc 1.96.0-beta.5 (a5a9a5438 2026-05-01)
binary: rustc
commit-hash: a5a9a5438ba8f6b83523031de99b15c12b44cea4
commit-date: 2026-05-01
host: x86_64-unknown-linux-gnu
release: 1.96.0-beta.5
LLVM version: 22.1.2
```

**Comment by samueltardieu** [comment]
Thanks for the report.

> I suspect that this is because the method is in a crate that is not a dependency of every crate in our worksapce?

That's it. Crate `tor-geoip` doesn't depend on `rand`. The solution here is:

- Create a different `clippy.toml` configuration file for every crate, with valid entries.
- Follow the suggestion (which has been created for this case, and for optional features), and add `allow-invalid = true` to the entry to indicate that you know that this dependency may be absent.

Does that makes sense to you? If it does, we can close the issue.

**Comment by ijackson** [comment]
I'm afraid that neither of these are good options.

> Create a different `clippy.toml` configuration file for every crate, with valid entries.

We have about 70 crates.  70 clippy.toml files containing near-identical contents would be a maintenance nightmare.

> Follow the suggestion (which has been created for this case, and for optional features), and add `allow-invalid = true` to the entry to indicate that you know that this dependency may be absent.

We have done this as a workaround, of course.  However, the "does this method actually exist" method is valuable and we would like to retain it.

This approach would mean that if the disallowed method *is* removed or renamed, nothing will notice.  Likewise if a slip means that the wrong method name is put in the list.

### Suggested solution

 * Suppress the warning if the whole crate in the path does not exist, in this depending crate.
 * Add a warning if there are entries with crate names that are not mentioned anywhere in the workspace.

### Other possible solutions

 1. Somehow collate all these warnings and intersect them at the end.  This seems very complicated and prone to getting things wrong when dependencies are conditional, etc.

 2. `allow-invalid-except-in = "crate-in-workspace"`

**Comment by ijackson** [comment]
> That's it. Crate `tor-geoip` doesn't depend on `rand`. The solution here is:

Are you sure this is everything that is going on?  Our current code after my workaround in https://gitlab.torproject.org/tpo/core/arti/-/merge_requests/3982 is this

```
    { path = "rand::random_range", reason = "Prefer tor_basic_utils::RngExt::gen_range_checked to avoid uncontrolled panics, or gen_range_infallible if applicable" },
    { path = "rand::RngExt::random_range", allow-invalid = true, reason = "Prefer tor_basic_utils::RngExt::gen_range_checked to avoid uncontrolled panics, or gen_range_infallible if applicable" },
```

Note that the disallow of `rand::random_range` is also in `rand`, but apparently does not require `allow-invalid`

So *something* more complicated must be going on that means it doesn't emit the warning.  Perhaps that logic has a bug which means it's not effective for the trait method?
